lib/elixir/lib/keyword.ex

```diff
@@ -1166,6 +1166,49 @@ defmodule Keyword do
     {:lists.reverse(take), :lists.reverse(drop)}
   end
 
+  @doc """
+  Splits the `keywords` into two keyword lists according to the given function
+  `fun`.
+
+  Splits the given `keywords` into two keyword lists by calling the provided
+  `fun` witch receives each `{key, value}` pair in the `keywords` as its only
+  argument. Returns a tuple with the first keyword list containing all the
+  elements in `keywords` for which applying `fun` returned a truthy value, and
+  a second keyword list with all the elements for which applying `fun` returned
+  a falsy value (`false` or `nil`).
+
+  See the examples below.
+
+  ## Examples
+
+      iex> Keyword.split_with([a: 1, b: 2, c: 3], fn {_k, v} -> rem(v, 2) == 0 end)
+      {[b: 2], [a: 1, c: 3]}
+
+      iex> Keyword.split_with([a: 1, b: 2, c: 3, b: 4], fn {_k, v} -> rem(v, 2) == 0 end)
+      {[b: 2, b: 4], [a: 1, c: 3]}
+
+      iex> Keyword.split_with([a: 1, b: 2, c: 3, b: 4], fn {k, v} -> k in [:a, :c] and rem(v, 2) == 0 end)
+      {[], [a: 1, b: 2, c: 3, b: 4]}
+
+      iex> Keyword.split_with([], fn {_k, v} -> rem(v, 2) == 0 end)
+      {[], []}
+
+  """
+  @doc since: "1.14.0"
+  @spec split_with(t, ({key, value} -> as_boolean(term))) :: {t, t}
+  def split_with(keywords, fun) when is_list(keywords) and is_function(fun, 1) do
+    fun = fn key_vaue_pair, {while_true, while_false} ->
+      if fun.(key_vaue_pair) do
+        {[key_vaue_pair | while_true], while_false}
+      else
+        {while_true, [key_vaue_pair | while_false]}
+      end
+    end
+
+    {while_true, while_false} = :lists.foldl(fun, {[], []}, keywords)
+    {:lists.reverse(while_true), :lists.reverse(while_false)}
```

Comment:
As there are no guarantees regarding keyword list ordering, I think this reverse isn't necessary.

Reply:
While its true that the ordering is not guaranteed - many operations, including `split/2` do retain the ordering, so I would be worried that the lack of consistency would be more troublesome than the actual change to the order of the elements.

I would be happy to drop the reverse, just wanted to make note of the reasoning that I included it!

Reply:
Correct! We can just return the result of foldl!

Reply:
@camilleryr we were writing at the same time. Actually, you are right, we need to reverse it, otherwise we affect the result of operations like `Keyword.get_values`. So please do keep the reverse!

Reply:
@josevalim Couldn't the reverse be omitted if `:lists.foldr` was used instead? Is there a performance aspect which makes `:lists.foldl` + reversing better than just `:lists.foldr`?

Reply:
Good question. foldr is just doing a reverse before folding, so at the end the result is the same.  :)

Reply:
Since they are the same - Ill swap it over to `foldr` so I can drop the extra line of code.  I also assume doing a single reverse might be more efficient than doing two reverses after the list has been split?